File: api/src/main/java/xtdb/RuntimeException.java
package xtdb;

import clojure.lang.IExceptionInfo;
import clojure.lang.IPersistentMap;
import clojure.lang.Keyword;

import java.util.Objects;

@SuppressWarnings("unused")
public class RuntimeException extends java.lang.RuntimeException implements IExceptionInfo {
    private static final long serialVersionUID = 7375922113394623815L;
    private static final Keyword RUNTIME_EXCEPTION = Keyword.intern("runtime-error");
    private static final Keyword ERROR_TYPE = Keyword.intern("xtdb.error", "error-type");
    private static final Keyword ERROR_KEY = Keyword.intern("xtdb.error","error-key");
    private static final Keyword MESSAGE_KEY = Keyword.intern("xtdb.error", "message");

    private final IPersistentMap data;

    public RuntimeException(String message, IPersistentMap data, Throwable cause) {
        super(message, cause);
        this.data = data;
    }

    @Override
    public IPersistentMap getData() {
        return data;
    }

    public static RuntimeException create(Keyword error, String message, IPersistentMap data) {
        return create(error, message, data,null);
    }
    public static RuntimeException create(Keyword error, String message, IPersistentMap data, Throwable cause) {
        message = message != null ? message : (String) data.valAt(MESSAGE_KEY, String.format("Runtime error: '%s'", error.toString()));
        return new RuntimeException(message, data.assoc(ERROR_TYPE, RUNTIME_EXCEPTION).assoc(ERROR_KEY, error).assoc(MESSAGE_KEY, message), cause);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        var that = (RuntimeException) o;
        return Objects.equals(getMessage(), that.getMessage()) && Objects.equals(getCause(), that.getCause()) && Objects.equals(data, that.data);
    }

    @Override
    public int hashCode() {
        return Objects.hash(getMessage(), getCause(), data);
    }

    @Override
    public String toString() {
        return String.format("RuntimeException{message=%s, data=%s}", getMessage(), data);
    }
}

File: api/src/main/kotlin/xtdb/RuntimeException.kt
package xtdb

import clojure.lang.IExceptionInfo
import clojure.lang.IPersistentMap
import clojure.lang.Keyword
import clojure.lang.PersistentHashMap
import clojure.lang.Util

private val ERROR_KEY: Keyword = Keyword.intern("xtdb.error", "error-key")

/**
 * @suppress
 */
@Suppress("unused")
data class RuntimeException(
    val key: Keyword?,
    override val message: String? = "Runtime error: '${key?.sym}'",
    private val data: Map<*, *> = emptyMap<Keyword, Any>(),
    override val cause: Throwable? = null,
) : java.lang.RuntimeException(message, cause), IExceptionInfo {

    override fun getData(): IPersistentMap =
        (data as? IPersistentMap ?: PersistentHashMap.create(data)).assoc(ERROR_KEY, key)

    // exclude `cause`
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RuntimeException) return false

        if (key != other.key) return false
        if (message != other.message) return false
        if (!Util.equiv(data, other.data)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = key?.hashCode() ?: 0
        result = 31 * result + (message?.hashCode() ?: 0)
        result = 31 * result + Util.hash(data)
        return result
    }
}
